### PR TITLE
bugfix: Fixed test suite discovery with the presence of companion object.

### PR DESCRIPTION
**Pull request by xydrolase**
## Background
This PR attempts to fix the bug reported in #5029, which causes unreliable test case discovery when companion objects for test classes exist.

The issue #5029 documents some of the behavior and some root-cause analysis. In short, the [`TestSuiteProvider.getTestEntries`](https://github.com/scalameta/metals/blob/34b3cc8f38792be4170344bb4bb685ba15dc3c2a/metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala#L412-L445) method has a caching mechanism which caches test cases computed for a specific symbol based on its **`fullyQualifiedName`**. However, because a test class (_e.g. `class FooSpec`) and its companion object (`object FooSpec`) have **exactly the same `fullyQualifiedName`), the clashing fully qualified symbol name can lead to flaky discovery: if the companion object happens to be processed **before** the test class itself, no test cases under the test class will be discovered.

## Solution Details
There are several ways to address this issue:

 1. Change how `fullyQualifiedName` is derived. For example, use `FooSpec$` to represent the fully qualified name of a companion object. However, because the method ([`BuildTargetClasses.symbolFromClassName`](https://github.com/scalameta/metals/blob/34b3cc8f38792be4170344bb4bb685ba15dc3c2a/metals/src/main/scala/scala/meta/internal/metals/debug/BuildTargetClasses.scala#L158-L174) used to compute `fullyQualifiedName` from symbols are used in multiple other places, changing this method breaks a lot of existing unit tests.
 2. Add a heuristic so that we make sure that the test class is always processed before its companion object. That is, if the developer defines both `class FooSpec extends AnyWordSpec` and `object FooSpec`, we want to make sure that we always cache the test cases discovered for `class FooSpec` as opposed to the latter (which should contain no test cases).

In this PR, we adopted solution 2. And to make sure that `class FooSpec` is processed before `object FooSpec`, we perform a lexical sort on all test symbols. This works because classes have suffix `#`, and objects have suffix `.`, based on the [SemanticDB specification](https://scalameta.org/docs/semanticdb/specification.html#symbol). Lexically, `#` is smaller than `.`, therefore, by sorting `currentTarget.testSymbols` by the symbol names, we can ensure that `class FooSpec` (which has symbol `FooSpec#`) comes before `object FooSpec` (which has symbol `FooSpec.`).

A unit test was added to verify that test cases can be discovered in the presence of the companion object.

## Results
Verified that with this fix, we can now consistently discover test cases in nvim-metals for ScalaTest when companion objects are present.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala [hunk 1]
```diff
@@ -419,6 +419,11 @@ final class TestSuitesProvider(
       currentTarget.testSymbols
         .readOnlySnapshot()
         .toList
+        // sort the symbols lexically so that symbols with the same fullyQualifiedName
+        // will be grouped, and the class will come before companion object (i.e.
+        // `a.b.WordSpec#` < `a.b.WordSpec.`). This ensures that the class is put into the cache
+        // instead of the companion object.
+        .sortBy(_._1)
```
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala:426]: Can you change it to `sortBy { case (symbol, _) => symbol }`?

**Review by kpodsiad — CHANGES_REQUESTED** [review]
Thanks for creating a PR @xydrolase! One nitpick, but otherwise PR is good to be merged.

File: metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala (hunk 1 above)
Comment (xydrolase) [reply metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala:426]: @kpodsiad Thanks for the review! I'll update the PR based on the comment.

**Comment by kpodsiad** [comment]
@xydrolase you can use snapshot version of metals if you don't want to wait untill next release to have that fix - grab latest one from https://scalameta.org/metals/docs

**Comment by xydrolase** [comment]
> @xydrolase you can use snapshot version of metals if you don't want to wait untill next release to have that fix - grab latest one from https://scalameta.org/metals/docs

Thanks for the tip!

**Pull request merged**